app/src/main/java/com/github/sdpcoachme/database/CachingStore.kt

```diff
@@ -629,6 +649,29 @@ class CachingStore(private val wrappedDatabase: Database,
         wrappedDatabase.removeChatListener(chatId)
     }
 
+    /**
+     * If the device is online, creates a WeatherPresenter and launch the weather forecast pipeline.
+     * Cached forecast is updated if the request completes normally.
+     * If not, simply returns the cached weather forecast.
+     *
+     * @param target The target location for the weather forecast
+     */
+    fun getWeatherForecast(target: LatLng): CompletableFuture<MutableState<WeatherForecast>> {
+
+        return if (isOnline()) {
+            // The WeatherPresenter will launch a weather request and return a future that if
+            // completed normally will update the cache.
+            val weatherPresenter = WeatherPresenter().bind(OpenMeteoRepository(RetrofitClient.api))
+            weatherPresenter.getWeatherForecast(target.latitude, target.longitude).thenApply {
+                weatherForecast = it
+                storeLocalData()
```

Comment:
by doing a `thenCompose { weatherForecast = it; storeLocalData() }.thenApply { weatherPresenter.observableWeatherForecast }` we can make sure the returned future always completes after the mutablestate has been updated. This could be useful for testing for example.

Reply:
Yes, this is way to include the coroutine job completion into a future completion. But then I don't know how to handle the view recomposition on future completion.